### PR TITLE
[IOTDB-3656] Loader Tool for TsFile

### PR DESCRIPTION
**Pull request by yschengzi**
add module for loading tsfile into IoTDB.
add tool script:

- load-tsfile.bat
- load-tsfile.sh

usage:
./load-tsfile.sh(load-tsfile.bat if Windows) -f <file> [-h <host>]
       [-help] [-p <port>] [-pw <password>] -u <username>
 -f <file>        File or Dictionary to be loaded.
 -h <host>        Host Name (optional, default 127.0.0.1)
 -help            Display help information(optional)
 -p <port>        Port (optional, default 6667)
 -pw <password>   password (optional)
 -u <username>    User name (required)

e.g.
./load-tsfile.bat -h 127.0.0.1 -p 6667 -u root -pw root -f "/home/tsfile"

User guide(Chinese):
https://apache-iotdb.feishu.cn/docx/doxcntEPZnXTGtQcdTk83yuOSqh

running example:
./load-tsfile.bat -h 127.0.0.1 -p 6667 -u root -pw root -f "D:\Pro\IoTDB\data\load\load"
Connect to IoTDB 127.0.0.1:6667 successfully.
349  [main] WARN  o.a.i.t.c.conf.TSFileDescriptor - not found iotdb-datanode.properties, use the default configs. 
Collect TsFiles successfully, 3 files to be loaded.
Start Loading TsFiles...%n
Loading D:\Pro\IoTDB\data\load\load\1655902209941-1-0-0.tsfile(1/3)...Done
Loading D:\Pro\IoTDB\data\load\load\1655902288903-2-0-0.tsfile(2/3)...Done
Loading D:\Pro\IoTDB\data\load\load\1655902340033-3-0-0.tsfile(3/3)...Done

### CODE REVIEW

File: load-tsfile-tool/src/main/java/org/apache/iotdb/RewriteFileTool.java [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+      } else {
+        printBoth(String.format("Start moving %s to backup dir.", filename));
+        session.executeNonQueryStatement(String.format("move '%s' '%s'", filename, backUpDirPath));
+      }
+      printBoth(String.format("Finish unloading %s.", filename));
+
+      // try to rewriteFile
+      printBoth(String.format("Start rewriting %s to iotdb.", filename));
+      if (targetFile.exists()) {
+        rewriteWrongTsFile(targetFilePath, session);
```
Comment (qiaojialin) [inline load-tsfile-tool/src/main/java/org/apache/iotdb/RewriteFileTool.java:166]:
```suggestion
        rewriteTsFile(targetFilePath, session);
```

File: load-tsfile-tool/src/main/java/org/apache/iotdb/RewriteFileTool.java [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    String badFilePath;
+    while ((badFilePath = bufferedReader.readLine()) != null) {
+      unloadAndReWriteWrongTsFile(badFilePath, session);
+    }
+    bufferedReader.close();
+    session.close();
+    printBoth("Finish rewriting all bad files.");
+  }
+
+  public static void unloadAndReWriteWrongTsFile(String filename, Session session) {
```
Comment (qiaojialin) [inline load-tsfile-tool/src/main/java/org/apache/iotdb/RewriteFileTool.java:145]: The TsFile to be rewrited is not managed by IoTDB, no need to unload.

File: load-tsfile-tool/src/main/java/org/apache/iotdb/RewriteFileTool.java [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+  private static final FSFactory fsFactory = FSFactoryProducer.getFSFactory();
+  private static final long MAX_TABLET_SIZE = 1024 * 1024;
+
+  private static PrintWriter writer;
+
+  /**
+   * -b=[path of backUp directory] -vf=[path of validation file]/-f=[path of tsfile list] -o=[path
+   * of output log] -u=[username, default="root"] -pw=[password, default="root"]
+   */
+  public static void main(String[] args) throws IOException {
```
Comment (qiaojialin) [inline load-tsfile-tool/src/main/java/org/apache/iotdb/RewriteFileTool.java:91]: This tool should input a folder containing TsFiles, or just one TsFile, then you rewrite them to IoTDB. 
No more file list file.

**Review by qiaojialin — CHANGES_REQUESTED** [review]

File: load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      } else if (listFile.getName().endsWith(TsFileConstant.TSFILE_SUFFIX)) {
+        list.add(listFile);
+      }
+    }
+    return list;
+  }
+
+  public static void writeToIoTDB(List<File> files, Session session) {
+    for (File file : files) {
+      RewriteFileTool.rewriteWrongTsFile(file.getPath(), session);
```
Comment (mychaow) [inline load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java:146]: if some file is corrupt, we need to rewrite all tsfiles again? Could we log it and support restart from the failed file?

File: load-tsfile-tool/src/main/java/org/apache/iotdb/RewriteFileTool.java (hunk 1 above)
Comment (yschengzi) [reply load-tsfile-tool/src/main/java/org/apache/iotdb/RewriteFileTool.java:166]: I delete all rewriteFile related.

File: load-tsfile-tool/src/main/java/org/apache/iotdb/RewriteFileTool.java (hunk 2 above)
Comment (yschengzi) [reply load-tsfile-tool/src/main/java/org/apache/iotdb/RewriteFileTool.java:145]: I delete all rewriteFile related.

File: load-tsfile-tool/src/main/java/org/apache/iotdb/RewriteFileTool.java (hunk 3 above)
Comment (yschengzi) [reply load-tsfile-tool/src/main/java/org/apache/iotdb/RewriteFileTool.java:91]: I delete all rewriteFile related.

File: load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java (hunk 4 above)
Comment (yschengzi) [reply load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java:146]: you can find one of solution in [user guide](https://apache-iotdb.feishu.cn/docx/doxcntEPZnXTGtQcdTk83yuOSqh), i give the list of tsfile which are already loaded, so people can find the tsfiles which do not need to load again.

File: load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+  private static String host = "localhost";
+  private static String port = "6667";
+  private static String user = "root";
+  private static String password = "root";
+  private static String filePath = "";
+
+  public static void main(String[] args) {
+    try {
+      parseArgs(args);
+      Session session = new Session(host, port, user, password);
```
[the remark this answers is not in the document]
Comment (yschengzi) [reply load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java]: done.

**Review by mychaow — APPROVED** [review]
LGTM

File: load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+          AbstractFileSeriesReader seriesReader;
+          if (chunkMetadataList.isEmpty()) {
+            seriesReader = new EmptyFileSeriesReader();
+            dataTypes.add(metadataQuerier.getDataType(path));
+          } else {
+            seriesReader = new FileSeriesReader(chunkLoader, chunkMetadataList, null);
+            dataTypes.add(chunkMetadataList.get(0).getDataType());
+          }
+          readersOfSelectedSeries.add(seriesReader);
+        }
```
Comment (JackieTien97) [inline load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java:295]: It seems that you read each sub sensor of an aligned device seperately, but it's better and more efficient to read them together.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+    List<File> unloadTsFiles = new ArrayList<>();
+    System.out.printf("Collect TsFiles successfully, %d files to be loaded.%n", size);
+    System.out.println("Start Loading TsFiles...");
+    for (int i = 0; i < size; i++) {
+      File file = files.get(i);
+      System.out.printf("Loading %s(%d/%d)...", file.getPath(), i + 1, size);
+      try {
+        writeTsFile(file.getPath(), session);
+      } catch (Exception e) {
+        System.out.println("Error");
```
[the remark this answers is not in the document]
Comment (yschengzi) [reply load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java:214]: done.

File: load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+        if (isAligned) {
+          session.insertAlignedTablet(tablet);
+        } else {
+          session.insertTablet(tablet);
+        }
+      }
+    }
+  }
+
+  private static void parseDeviceFromTsFile(String filename) throws IOException {
```
[the remark this answers is not in the document]
Comment (yschengzi) [reply load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java:328]: ChunkMetadata doesn't contain Encoding and CompressionType of measurmentSchema, can not contract measurementSchema from ChunkMetadata.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/reader/page/TimePageReader.java [hunk 9]
```diff
@@ -80,7 +83,10 @@ public long[] getNextTimeBatch() throws IOException {
     } else {
       List<Long> timeList = new ArrayList<>();
       while (timeDecoder.hasNext(timeBuffer)) {
-        timeList.add(timeDecoder.readLong(timeBuffer));
+        long timestamp = timeDecoder.readLong(timeBuffer);
+        if (!isDeleted(timestamp)) {
+          timeList.add(timestamp);
+        }
```
[the remark this answers is not in the document]
Comment (yschengzi) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/read/reader/page/TimePageReader.java:89]: done.

File: load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+            if (field == null) {
+              tablet.bitMaps[i].mark(tablet.rowSize);
+            } else {
+              tablet.addValue(
+                  measurementSchemas.get(i).getMeasurementId(),
+                  tablet.rowSize,
+                  field.getObjectValue(field.getDataType()));
+            }
+          }
+          tablet.rowSize++;
```
[the remark this answers is not in the document]
Comment (yschengzi) [reply load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java:317]: done.

File: load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java (hunk 6 above)
Comment (yschengzi) [reply load-tsfile-tool/src/main/java/org/apache/iotdb/TsFileLoaderTool.java:294]: Actually,
```
QueryDataSet dataSet =
            new DataSetWithoutTimeGenerator(paths, dataTypes, readersOfSelectedSeries);
```
is the real query statement, so aligned timeseries is queried by aligned API.

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**